nltk/translate/chrf_score.py

```diff
@@ -110,28 +132,59 @@ def corpus_chrf(list_of_references, hypotheses, min_len=1, max_len=6, beta=3.0):
 
     # Iterate through each hypothesis and their corresponding references.
-    for reference, hypothesis in zip(list_of_references, hypotheses):
-        # Cheating condition to allow users to input strings instead of tokens.
-        if type(reference) and type(hypothesis) != str:
-            reference, hypothesis = ' '.join(reference), ' '.join(hypothesis)
-        # For each order of ngram, calculate the no. of ngram matches and
-        # keep track of no. of ngram in references.
-        ref_ngrams = Counter(everygrams(reference, min_len, max_len))
-        hyp_ngrams = Counter(everygrams(hypothesis, min_len, max_len))
-        overlap_ngrams = ref_ngrams & hyp_ngrams
-        tp = sum(overlap_ngrams.values()) # True positives.
-        tpfp = sum(hyp_ngrams.values()) # True positives + False positives.
-        tffn = sum(ref_ngrams.values()) # True posities + False negatives.
-
-    precision = tp / tpfp
-    recall = tp / tffn
-    factor = beta**2
-    score = (1+ factor ) * (precision * recall) / ( factor * precision + recall)
-    return score
+    for reference, hypothesis in zip(references, hypotheses):
+        num_sents += 1
+
+        # preprocess both reference and hypothesis
+        reference = _preprocess(reference, ignore_whitespace)
+        hypothesis = _preprocess(hypothesis, ignore_whitespace)
+
+        # Calculate f-scores for each sentence and for each n-gram order
+        # separately.
+        for order in range(min_len, max_len + 1):
+            ref_ngrams = Counter(ngrams(reference, order))
+            hyp_ngrams = Counter(ngrams(hypothesis, order))
+
+            # calculate the number of ngram matches
+            overlap_ngrams = ref_ngrams & hyp_ngrams
+            tp = sum(overlap_ngrams.values())  # True positives.
+            tpfp = sum(hyp_ngrams.values())  # True positives + False positives.
+            tffn = sum(ref_ngrams.values())  # True posities + False negatives.
+
+            # Avoid division by zero errors. We also return 0 when both
+            # reference and hypothesis are empty.
+            if 0 in (tp, tpfp, tffn):
+                n_gram_fscores[order].append(0.0)
+            else:
+                prec = tp / tpfp  # precision
+                rec = tp / tffn  # recall
+                factor = beta**2
+                f_score = (1 + factor) * (prec * rec) / (factor * prec + rec)
+                n_gram_fscores[order].append(f_score)
+
+    # This is not specified in the paper but the author's implementation
+    # computes macro-averages both over n-gram lengths and sentences.
+
+    # how many n-gram sizes
+    num_ngram_sizes = max_len + 1 - min_len
+
+    # sum of f-scores over all sentences for each n-gram order
+    total_scores = [sum(n_gram_fscores[o]) for o in range(min_len, max_len + 1)]
+
+    # macro-average over n-gram orders and over all sentences
+    return (sum(total_scores) / num_ngram_sizes) / num_sents
```

Comment:
Regarding macro vs micro average, we can do better than the original implementation by allowing a parameter with `average="macro"` as default. That'll follow the API design of `sklearn` metrics.

Reply:
Hmmm, looking at https://github.com/m-popovic/chrF/blob/master/chrF%2B%2B.py#L186 

It's actually taking the micro average not macro-average across all hypotheses. 
For each hypothesis, it's not taking average across all references but the argmax.

Reply:
Right: across **references**, we take the **maximum** F-score.

Across **n-gram orders**, we always do **macro** average.

Now across **sentences**, the code outputs both macro and micro average, I missed that. This is the output for the test corpus of `corpus_chrf`:

```
$ python chrF++.py -R tests.references.txt -H tests.hypotheses.txt -nw 0 -b 3
start_time:	1534343254
c6+w0-F3	38.1372
c6+w0-avgF3	39.1009
end_time:	1534343254
```

So, `avgF3` is the macro average (what I implemented) and `F3` is the micro average. So I think it would make sense to add the parameter and perhaps make "micro" the default value.

I assume it should only apply to averaging over sentences in the corpus, do you agree?

Reply:
Yes, I think `micro` and `marco` needs to be a parameter. 
I think Maja's code follows the old perl style BLEU and NIST code that's why she was keeping hashmaps and then using different averaging techniques based on what's easiest to access from the hashmap.

I'll take a look at this again tomorrow morning (~10-12 hours) later and see how we can work this out elegantly.